### ISSUE TITLE
Wrong cast being generated for H2's MySQL style enum types, when a EnumConverter is present

### ISSUE THREAD

**Issue by djarnis73** [jOOQ/jOOQ#17019]
### Expected behavior

H2 allows for the following create table syntax:

```
create table sample (
    dialect1 enum ('MYSQL','MARIADB') not null
);
```
When you use the jooq code generator to map this using an enumconverter like:

```
<database>
    <forcedTypes>
        <forcedType>
            <enumConverter>true</enumConverter>
            <userType>org.jooq.SQLDialect</userType>
            <includeExpression>dialect.</includeExpression>
        </forcedType>
    </forcedTypes>
</database>
```
It will create a table field like:
```
public final TableField<TestRecord, SQLDialect> DIALECT1 = createField(DSL.name("DIALECT1"), DefaultDataType.getDefaultDataType("\"MCVE\".\"TEST_DIALECT1\"").nullable(false), this, "", new EnumConverter<Object, SQLDialect>(Object.class, SQLDialect.class));
```
Note the datatype value: `DefaultDataType.getDefaultDataType("\"MCVE\".\"TEST_DIALECT1\"")` which looks like the column name.

If you instead create the table like (without using the enum syntax)
```
create table sample (
  dialect2 varchar(255) not null check (dialect2 in ('MYSQL','MARIADB'))
);
```
the fields gets created like:
```
public final TableField<TestRecord, SQLDialect> DIALECT2 = createField(DSL.name("DIALECT2"), SQLDataType.VARCHAR(255).nullable(false), this, "", new EnumConverter<String, SQLDialect>(String.class, SQLDialect.class));
```
note the fixed datatype: `SQLDataType.VARCHAR(255)`

### Actual behavior

I would expect jooq to infer the varchar datatype.

### Steps to reproduce the problem

I have created an mcve here: https://github.com/djarnis73/jOOQ-mcve where i have modified the jOOQ-mcve-java-h2 module (maven).

### jOOQ Version

Verified on 3.18.X, 3.19.x probably also others

### Database product and version

H2 2.2.224

### Java Version

openjdk version "21.0.2" 2024-01-16

**Comment by djarnis73** [comment]
I'm not sure this is H2 specific or a general bug for databases supporting enum column syntax.

**Comment by djarnis73** [comment]
Hi @lukaseder 

Thanks for your response. I guess I did not really mention it, but my problem is that it leads to sql syntax error because jooq generates sql like:

```
select DIALECT1, DIALECT2 from TEST where DIALECT1 = cast(? as "MCVE"."TEST_DIALECT1")
```

Where `"MCVE"."TEST_DIALECT1"` is not a valid datatype.

So I believe a safer default datatype for enums would be varchar(255).

You can try it out in my reproducer project (just run mvn test in the jOOQ-mcve-java-h2 module)

Best regards Jens

**Comment by djarnis73** [comment]
I also noticed that if you run the generator without the enumConverter, then it generates a varchar datatype for the column.

**Comment by lukaseder** [comment]
Thanks for clarifying. I'll look into the reproducer later today

**Comment by lukaseder** [comment]
This is a bind value specific cast. As a workaround, you could just inline the value using `DSL.inline()`, or auto-inlining (if you're licensed):
https://www.jooq.org/doc/latest/manual/sql-building/dsl-context/custom-settings/settings-auto-inline-bind-values/

**Comment by lukaseder** [comment]
The reason why we're not generating any `EnumType` values (and thus why we need the `DefaultDataType.getDefaultDataType` call) is this:

- https://github.com/jOOQ/jOOQ/issues/1137

Maybe, that very old change was done for a reason that was valid at the time (most probably generate less "unnecessary" code), but now, I'm not convinced it was really a good change. The `T` type should be the generated `EnumType`, and the `U` type is your own enum type, and the conversion should be between these two.

I'll check if we can revert #1137

**Comment by lukaseder** [comment]
Reverting #1137 would fix the problem here, but it would be an incompatible change as the conversion would now be between `T = SomeEnumType` and `U`, instead of `T = Object` and `U`. While in H2, the problem you've observed would have probably prevented people from using this feature combination, there might be 1-2 systems out there working with MySQL, where this isn't a problem (bind values don't need to be cast in MySQL).

**Comment by lukaseder** [comment]
I cannot think of any other fix right now that wouldn't involve complex handling of this edge case in the code generator (which can by itself introduce new edge cases and/or regressions).

Would it be OK if the thorough fix (reverting #1137) is applied only to 3.20.0 and you work with the workaround for the time being?

Another workaround could be to apply both data type rewriting (rewrite the type to `VARCHAR`) and your converter:
https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-forced-types/codegen-database-forced-types-rewrite/

**Comment by lukaseder** [comment]
Fixed in version 3.20.0

**Comment by djarnis73** [comment]
I have a usable workaround, so a fix for 3.20.x is fine by me.
